### PR TITLE
[WIP] Make AuthenticatorData immutable and strongly type throughout library

### PR DESCRIPTION
**Pull request by iamcarbon**
This PR prevents the mutation of AuthenticatorData after parsing, and strongly types it throughout the library.

Ready for review.

### CODE REVIEW

File: Src/Fido2/Objects/AuthenticatorData.cs [hunk 1]
```diff
@@ -61,103 +85,90 @@ public sealed class AuthenticatorData
...
+    private byte[]? _data = null;
 
-    /// <summary>
-    /// Attested credential data is a variable-length byte array added to the 
-    /// authenticator data when generating an attestation object for a given credential.
-    /// </summary>
-    public AttestedCredentialData AttestedCredentialData;
+    public byte[] ToByteArray()
+    {
+        if (_data != null)
```
Comment (iamcarbon) [inline Src/Fido2/Objects/AuthenticatorData.cs:102]: We cache the data, to avoid additional allocations after parsing.

**Review by abergs — APPROVED** [review]
So far, looks good. Is it still WIP?

File: Test/Fido2Tests.cs [hunk 2]
```diff
@@ -600,18 +586,15 @@ public async Task TestTPMSHA1AttestationAsync()
...
+
     [Fact]
     public async Task TestAndroidKeyAttestationAsync()
     {
         var jsonPost = JsonSerializer.Deserialize<AuthenticatorAttestationRawResponse>(await File.ReadAllTextAsync("./attestationAndroidKeyResponse.json"));
         var options = JsonSerializer.Deserialize<CredentialCreateOptions>(await File.ReadAllTextAsync("./attestationAndroidKeyOptions.json"));
         var o = AuthenticatorAttestationResponse.Parse(jsonPost);
         await o.VerifyAsync(options, _config, (x, cancellationToken) => Task.FromResult(true), _metadataService, CancellationToken.None);
-        byte[] ad = o.AttestationObject.AuthData;
-        // TODO : Why read ad ? Is the test finished ?
```
[the remark this answers is not in the document]
Comment (iamcarbon) [reply Test/Fido2Tests.cs:614]: Yep, this logic was a dead / unused.

File: Src/Fido2/AuthenticatorResponse.cs [hunk 3]
```diff
@@ -16,6 +14,14 @@ namespace Fido2NetLib;
 /// </summary>
 public class AuthenticatorResponse
 {
+    [JsonConstructor]
+    public AuthenticatorResponse(string type, byte[] challenge, string origin) // for deserialization
```
[the remark this answers is not in the document]
Comment (iamcarbon) [reply Src/Fido2/AuthenticatorResponse.cs:18]: Yep, System.Text.Json will use the [JsonConstructor] during serialization (both in reflection and source generator mode). I'll add some additional test coverage here.

[the remark this answers is not in the document]
Comment (iamcarbon) [reply Src/Fido2/AuthenticatorResponse.cs:18]: Added test coverage.

**Comment by iamcarbon** [comment]
@abergs Ready for review!

File: Src/Fido2/AuthenticatorAttestationResponse.cs [hunk 4]
```diff
@@ -346,6 +336,23 @@ public ParsedAttestationObject(string fmt, CborMap attStmt, byte[] authData)
         
         public CborMap AttStmt { get; }
 
-        public byte[] AuthData { get; }
+        public AuthenticatorData AuthData { get; }
+
+        internal static ParsedAttestationObject FromCbor(CborObject cbor)
```
Comment (iamcarbon) [inline Src/Fido2/AuthenticatorAttestationResponse.cs:341]: This method also eliminates double lookups.

**Review by abergs — APPROVED** [review]

**Pull request merged**